src/desktop/space.rs

```diff
@@ -0,0 +1,419 @@
+        damage.retain(|rect| rect.overlaps(output_geo));
+        damage.retain(|rect| rect.size.h > 0 && rect.size.w > 0);
+        for rect in damage.clone().iter() {
+            // if this rect was already removed, because it was smaller as another one,
+            // there is no reason to evaluate this.
+            if damage.contains(rect) {
+                // remove every rectangle that is contained in this rectangle
+                damage.retain(|other| !rect.contains_rect(*other));
+            }
+        }
+
+        let output_transform: Transform = output.current_transform().into();
+        if let Err(err) = renderer.render(
+            output_transform
+                .transform_size(output_size)
+                .to_f64()
+                .to_physical(state.render_scale)
+                .to_i32_round(),
+            output_transform,
+            |renderer, frame| {
+                // First clear all damaged regions
+                for geo in &damage {
+                    slog::debug!(self.logger, "Clearing at {:?}", geo);
+                    frame.clear(
+                        clear_color,
+                        Some(geo.to_f64().to_physical(state.render_scale).to_i32_ceil()),
+                    )?;
+                }
+
+                // Then re-draw all window overlapping with a damage rect.
+                for window in self.windows.iter() {
+                    let wgeo = window_rect(window, &self.id);
+                    let mut loc = window_loc(window, &self.id);
+                    if damage.iter().any(|geo| wgeo.overlaps(*geo)) {
+                        loc -= output_geo.loc;
+                        slog::debug!(self.logger, "Rendering window at {:?}", wgeo);
+                        draw_window(renderer, frame, window, state.render_scale, loc, &self.logger)?;
+                        window_state(self.id, window).drawn = true;
+                    }
+                }
+
+                Result::<(), R::Error>::Ok(())
+            },
+        ) {
+            // if the rendering errors on us, we need to be prepared, that this whole buffer was partially updated and thus now unusable.
+            // thus clean our old states before returning
+            state.old_damage = VecDeque::new();
+            state.last_state = IndexMap::new();
+            return Err(RenderError::Rendering(err));
+        }
+
+        // If rendering was successful capture the state and add the damage
+        state.last_state = self
+            .windows
+            .iter()
+            .map(|window| {
+                let wgeo = window_rect(window, &self.id);
+                (window.0.id, wgeo)
+            })
+            .collect();
+        state.old_damage.push_front(new_damage);
+
+        // Return if we actually rendered something
+        Ok(!damage.is_empty())
+    }
+
+    pub fn send_frames(&self, all: bool, time: u32) {
+        for window in self.windows.iter().filter(|w| {
+            all || {
```

Comment:
What's that?

Reply:
I though there are reasons to send frame events periodically to all clients, even those not mapped (the surface-suspension protocol wants to fix this, I think). So I did not want to force the user to do it either way.

Reply:
Ah no, I meant that `all || { ... }` syntax, is that some new rust feature? I've never seen it.

Reply:
that is just `boolean <OR> { expr returning a boolean }`. Not a closure, just a simple `||` you know like `&&`.

Reply:
Ah right, my brain is too hard-wired to read `| | {` as the start of a closure it appears. 😭